cmd/action/docker.go

```diff
@@ -0,0 +1,232 @@
+// Copyright 2021-present The Atlas Authors. All rights reserved.
+// This source code is licensed under the Apache 2.0 license found
+// in the LICENSE file in the root directory of this source tree.
+
+package action
+
+import (
+	"bytes"
+	"context"
+	"database/sql"
+	"errors"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"log"
+	"net"
+	"os"
+	"os/exec"
+	"strconv"
+	"strings"
+	"time"
+
+	mysqld "github.com/go-sql-driver/mysql"
+)
+
+const pass = "pass"
+
+type (
+	// DockerConfig is used to configure container creation.
+	DockerConfig struct {
+		Image string
+		Env   []string  // env vars to pass to the container
+		Port  string    // internal port to expose and connect to
+		Out   io.Writer // custom writer to send docker cli output to
+	}
+	// A Container is an instance of a created container.
+	Container struct {
+		cfg DockerConfig // DockerConfig used to create this container
+		// ID of the container.
+		ID string
+		// Passphrase of the root user.
+		Passphrase string
+		// Port on the host this containers service is bound to.
+		Port string
+	}
+	// ConfigOption allows configuring DockerConfig with functional arguments.
+	ConfigOption func(*DockerConfig) error
+)
+
+// NewConfig returns a new config with the given options applied.
+func NewConfig(opts ...ConfigOption) (*DockerConfig, error) {
+	c := &DockerConfig{Out: os.Stdout}
+	for _, opt := range opts {
+		if err := opt(c); err != nil {
+			return nil, err
+		}
+	}
+	return c, nil
+}
+
+// MySQL returns a new DockerConfig for a MySQL image.
+func MySQL(version string, opts ...ConfigOption) (*DockerConfig, error) {
+	return NewConfig(append(
+		[]ConfigOption{
+			Image("mysql:" + version),
+			Port("3306"),
+			Env("MYSQL_ROOT_PASSWORD=" + pass),
+		},
+		opts...,
+	)...,
+	)
+}
+
+// Image sets the docker image to use.
+//
+// Example:
+//   - Image("mysql")
+//   - Image("postgres:13")
+func Image(i string) ConfigOption {
+	return func(c *DockerConfig) error {
+		c.Image = i
+		return nil
+	}
+}
+
+// Port sets the port the container services exposes.
+//
+// Common configuration include:
+//   - mysql: Port("3306")
+//   - pg:    Port("5432")
+func Port(p string) ConfigOption {
+	return func(c *DockerConfig) error {
+		c.Port = p
+		return nil
+	}
+}
+
+// Env sets the environment variables to pass to the container.
+//
+// Example:
+//   - mysql: Env("MYSQL_ROOT_PASSWORD=password")
+//   - pg:    Env("POSTGRES_PASSWORD=password")
```

Comment:
Write examples in Go syntax.
```suggestion
// Env sets the environment variables to pass to the container. For example:
//
// 	Config(Image("mysql"), Env("MYSQL_ROOT_PASSWORD=password"))
// 	Config(Image("postgres"), Env("MYSQL_ROOT_PASSWORD=password"))
//
```